[Clang][Driver][Test] Created test for unsupported driver options

GeorgeKA opened the pull request:
Closes #119190 

Created _generate_unsupported_in_drivermode.py_ which generates Lit regression test that validate that options are only exposed to intended driver modes.

The options and driver modes are parsed from _Options.td_, whose path should be provided on the command line. See
_clang/include/clang/Driver/Options.td_

The path to the TableGen executable can optionally be provided. Otherwise, the script will search for it.

### Proposed workflow
- A change is made to _clang/include/clang/Driver/Options.td_
- Regenerate the Lit tests that validate exposure of options to intended driver modes
`python generate_unsupported_in_drivermode.py --llvm-bin <path>/llvm-project/build/bin --llvm-tblgen llvm-tblgen ../include/clang/Driver/Options.td`
- Run the Lit tests to check exposure                                                                                                                                                                           
`llvm-lit -v clang/test/Driver/unsupported_in_drivermode.c`
`llvm-lit -v clang/test/Driver/flang/unsupported_in_drivermode.c`
`llvm-lit -v flang/test/Driver/unsupported_in_flang_fc1.f90`

GeorgeKA committed:
[clang] Document the return value of __builtin_COLUMN

PR for issue #78657

Updated clang/docs/LanguageExtensions.rst to detail the return value of
__builtin_COLUMN for this implementation.

GeorgeKA committed:
Updated the documentation link to the standard instead of cppreference

GeorgeKA committed:
[Clang][Driver][Test] Created test for unsupported driver options

Created generate_unsupported_in_drivermode.py which generates a Lit
regression test file that validates that options are only exposed to
intended driver modes.

The options and driver modes are parsed from Options.td, whose path
should be provided on the command line. See
clang/include/clang/Driver/Options.td

The path to the TableGen executable can optionally be provided.
Otherwise, the script will search for it.

Maetveis commented:
Hi,Thanks for working on this issue. I wanted to preface this by feel free to ignore this during the holidays, zero pressure. I like to interact with the community in my free time, but that obviously does not need to extend to everyone :).

What kind of failures are you seeing in the generated test? There shouldn't be any afaict if the script works correctly, as Options.td IS the source where the parsing code for the various drivers are generated from. Are you perhaps missing the case where there is no visibility specified at all and it should be assumed to be `DefaultVisibility`, i.e. the gcc compatible clang driver?

GeorgeKA commented:
@Maetveis 

All of the Visibility fields appear to be not null, but that's a good check to add anyway. Thanks. After adding it though, the same failures occur. Here's an example. Maybe there's a gap in my understanding:

The script generates tests for both flang & clang -cc1as when looking at the O_flag option. cc1as outputs the unexpected error message, but flang does not.
```
$ clang -cc1as -O_flag -help 2>&1 | less
clang -cc1as: error: unknown argument: '-O_flag'

$ flang -O_flag -help 2>&1 | less
<regular help message>
```
These are the steps I'm taking for manual validation:
1) Generate JSON dump for Options.td
`llvm-tblgen -I <llvm repo>/llvm/include <llvm repo>/clang/include/clang/Driver/Options.td -dump-json | jq . > Options.td.json`
2) Navigate to `"O_flag": {`
3) "`Group`" is null, so no extra visibilities to check
4) "`Visibility`" includes DefaultVis, CC1Option, & FC1Option, which leaves unsupported Visibilities as CC1AsOption, CLOption, DXCOption, FlangOption, so the script generates run lines & checks for these four.

Drivers/visibilities include:
	"CC1AsOption"
	"CC1Option"
	"CLOption"
	"DXCOption"
	"DefaultVis"
	"FC1Option"
	"FlangOption"

GeorgeKA committed:
Fixed formatting and changed to common error msg check

Changed to a common error message for the various driver modes,
and formatted with Python darker.

Maetveis commented:
> @Maetveis
> 
> All of the Visibility fields appear to be not null, but that's a good check to add anyway. Thanks. After adding it though, the same failures occur. Here's an example. Maybe there's a gap in my understanding:
> 
> The script generates tests for both flang & clang -cc1as when looking at the O_flag option. cc1as outputs the unexpected error message, but flang does not.

That looks to be by design, because you are using the `-help`, the drivers meant to be user facing (clang, clang-cl and flang) don't report unknown options with `-help`. A user might tack on a `-help` after their broken command line when they are looking for the exact spelling of their option so it makes sense that it is not an error to have unknown options when mixed with `-help`.

Simply you shouldn't be using `-help`, instead passing options to perform a real compilation on the command line like `clang-cl <tested-options> -### -x c++ -c - < /dev/null` should work better.

The other issues is that `-O_flag` is parsed as `-O _flag` for drivers that support `-O`, in runtime longer valid options have preference over shorter ones, in this case I think if there is a supported joined option (`Kind == JoinedClass`) that is a prefix of an unsupported option you now way to test the unsupported option so it should be skipped.

GeorgeKA committed:
A variety of additions to more properly generate the Lit test.

- Added handling for false positives caused by supported options that are prefixes for unsupported ones.
- Added a controller to simplify modifying the test commands for each driver.
- Generally cleaned things up

- A list of exceptions called exceptions_sequence that needs to be fixed
- clang-cl & clang-dxc testing remains

GeorgeKA committed:
Generate the Lit test in clang/test/Options

GeorgeKA commented:
@Maetveis 
Do you have any thoughts on the following two groups of options that are unexpectedly succeeding?
I'll note that I did come across some options for which the driver appears to correct user error by adding the appropriate driver option, like testing "`-fexperimental-sanitize-metadata=atomics`" with the default driver and the driver adding `-cc1`, but I'm not aware of a supported `-cc1` option that outputs the final list of options used like `-###` does for the default driver.

1) "`fheinous-gnu-extensions`" & "`fcuda-approx-transcendentals`" with `-cc1`
Both of these unexpectedly succeed with `clang -cc1`. Notably, they both have aliases that are supported for `-cc1`, ([ex: fheinous-gnu-extensions](https://github.com/llvm/llvm-project/blob/75781fa3c77f14640459a97d4265dda76d80e453/clang/include/clang/Driver/Options.td#L2823)), but those aliases expectedly fail when used with `-cc1`, which suggests aliases aren't the issue.

Ex: `clang -cc1 -fheinous-gnu-extensions -x c++ - < /dev/null 2>&1`

2) "`mno-strict-align`" & "`mstrict-align`" with `-cc1`
Both of these unexpectedly succeed with `clang -cc1`. Both of these have no alias. They're part of the m_Group and the CompileOnly_Group, which don't include `-cc1` for visibility.

Ex: `clang -cc1 -mno-strict-align -x c++ - < /dev/null 2>&1`

Maetveis commented:
Sorry I was caught up with work, and didn't find the time to look at this until now.

> @Maetveis Do you have any thoughts on the following two groups of options that are unexpectedly succeeding? I'll note that I did come across some options for which the driver appears to correct user error by adding the appropriate driver option, like testing "`-fexperimental-sanitize-metadata=atomics`" with the default driver and the driver adding `-cc1`

For `-fexperimental-sanitize-metadata=atomics` I think it should be hidden by `-fexperimental-sanitize-metadata=<anything>` (Which is driver only). Maybe your search logic for hidden options has bugs, but I didn't dive into it to check.


> 1. "`fheinous-gnu-extensions`" & "`fcuda-approx-transcendentals`" with `-cc1`
>    Both of these unexpectedly succeed with `clang -cc1`. Notably, they both have aliases that are supported for `-cc1`, ([ex: fheinous-gnu-extensions](https://github.com/llvm/llvm-project/blob/75781fa3c77f14640459a97d4265dda76d80e453/clang/include/clang/Driver/Options.td#L2823)), but those aliases expectedly fail when used with `-cc1`, which suggests aliases aren't the issue.
> 
> Ex: `clang -cc1 -fheinous-gnu-extensions -x c++ - < /dev/null 2>&1`
> 
> 2. "`mno-strict-align`" & "`mstrict-align`" with `-cc1`
>    Both of these unexpectedly succeed with `clang -cc1`. Both of these have no alias. They're part of the m_Group and the CompileOnly_Group, which don't include `-cc1` for visibility.
> 
> Ex: `clang -cc1 -mno-strict-align -x c++ - < /dev/null 2>&1`

These all produce errors for me. Are you sure you're testing everything right? You have an up-to-date clang build for example? The testing tool should give you an output like `llvm-lit: (...)/llvm/utils/lit/lit/llvm/config.py:506: note: using clang (...)/build/bin/clang` that should be pointing into the llvm build folder.

GeorgeKA committed:
Merge branch 'llvm:main' into main

GeorgeKA committed:
Merge branch 'main' of github.com:GeorgeKA/llvm-project into 119190-unsupportedOptions

GeorgeKA committed:
Added clang-cl & clang-dxc logic

Additional changes to improve correctness, including preprocessing to
filter automatically added driver options, changed FileCheck strings,
etc.

The script generates two test files, since the flang tests are under the
flang directory.
    Driver/unsupported_in_drivermode.c
    Driver/flang/unsupported_in_flang.f90

Notably, As per flang.f90, "-fc1 is invoked when in
--driver-mode=flang", so tests were not generated for visibility
FlangOption; only for FC1Option.

GeorgeKA commented:
Ahh, thanks. You're right, I had environment issues. Now that Lit is pointing to an up to date clang, the exceptions I mentioned properly fail. That's a relief.

Fyi, I think all that's left before it's ready for review is handling for aliases. I'll try and add that later today or tomorrow.

GeorgeKA committed:
Added handling for aliases

More additions:
- Redirect stderr to stdout for dxc & cl mode.
- Pass/fail RUN line error handling due to unreliable return codes when
  preprocessing commands.

GeorgeKA committed:
Merge branch 'llvm:main' into main

GeorgeKA committed:
Merge branch 'main' of github.com:GeorgeKA/llvm-project into 119190-unsupportedOptions

GeorgeKA committed:
Removed the clang_ignored_f_Group comments

GeorgeKA committed:
Merge branch 'llvm:main' into 119190-unsupportedOptions

GeorgeKA committed:
Merge branch '119190-unsupportedOptions' of github.com:GeorgeKA/llvm-project into 119190-unsupportedOptions

GeorgeKA committed:
Cleanup from previous merge

GeorgeKA committed:
More merge issue fixes

GeorgeKA committed:
Disabled tests on Windows

GeorgeKA commented:
Hi @Maetveis . This is ready for review.

There are a few things that I came across while making the script that I should highlight :

1) clang -cc1 test directory

[test/Driver/lit.local.cfg](https://github.com/llvm/llvm-project/blob/e10d551aa482ee185a80216b2670a2947a8bdeb0/clang/test/Driver/lit.local.cfg#L27) says, '_("%clang_cc1", """*** Do not use 'clang -cc1' in Driver tests. ***""")_', yet I see various tests in that directory that do so, like [linker-wrapper.c](https://github.com/llvm/llvm-project/blob/e10d551aa482ee185a80216b2670a2947a8bdeb0/clang/test/Driver/linker-wrapper.c#L10). If there's no opposition, I'll leave the cc1 tests in the main Lit test file, Driver/unsupported_in_drivermode.c.

2) flang vs flang -fc1

As per [clang/test/Driver/flang/flang.f90](https://github.com/llvm/llvm-project/blob/e10d551aa482ee185a80216b2670a2947a8bdeb0/clang/test/Driver/flang/flang.f90#L1), "flang -fc1 is invoked when in --driver-mode=flang", so I disabled the plain flang tests since there's no distinction.

3) Testing on Windows

I disabled the tests on Windows given the following recursion error that occurs on Windows regression machines. (It has some odd formatting. Pasting in a text editor makes viewing it easier.)
```
	^[_bk;t=1737572755765^GPASS: Clang :: Driver/fsanitize-memory-param-retval.c (11017 of 21306)^M                                                                                                             ^[_bk;t=1737572756104^G******************** TEST 'Clang :: Driver/flang/unsupported_in_flang.f90' FAILED ********************^M                                                                             ^[_bk;t=1737572756104^G******************** TEST 'Clang :: Driver/flang/unsupported_in_flang.f90' FAILED ********************^M                                                                             ^[_bk;t=1737572756104^GTraceback (most recent call last):^M                                                                                                                                                 ^[_bk;t=1737572756104^GTraceback (most recent call last):^M                                                                                                                                                 ^[_bk;t=1737572756104^GTraceback (most recent call last):^M                                                                                                                                                 ^[_bk;t=1737572756104^G  File "C:\ws\src\llvm\utils\lit\lit\formats\shtest.py", line 29, in execute^M
	^[_bk;t=1737572756104^G    return lit.TestRunner.executeShTest(^M
	^[_bk;t=1737572756104^G  File "C:\ws\src\llvm\utils\lit\lit\TestRunner.py", line 2326, in executeShTest^M                                                                                                   ^[_bk;t=1737572756104^G  File "C:\ws\src\llvm\utils\lit\lit\TestRunner.py", line 2326, in executeShTest^M                                                                                                   ^[_bk;t=1737572756104^G  File "C:\ws\src\llvm\utils\lit\lit\TestRunner.py", line 2326, in executeShTest^M                                                                                                   ^[_bk;t=1737572756104^G  File "C:\ws\src\llvm\utils\lit\lit\TestRunner.py", line 2326, in executeShTest^M                                                                                                   ^[_bk;t=1737572756104^G  File "C:\ws\src\llvm\utils\lit\lit\TestRunner.py", line 2326, in executeShTest^M                                                                                                   ^[_bk;t=1737572756104^G    res = executeScriptInternal(^M
	^[_bk;t=1737572756104^G  File "C:\ws\src\llvm\utils\lit\lit\TestRunner.py", line 1110, in executeScriptInternal^M
	^[_bk;t=1737572756104^G    exitCode, timeoutInfo = executeShCmd(^M                                                                                                                                          ^[_bk;t=1737572756104^G    exitCode, timeoutInfo = executeShCmd(^M                                                                                                                                          ^[_bk;t=1737572756104^G    exitCode, timeoutInfo = executeShCmd(^M                                                                                                                                          ^[_bk;t=1737572756104^G    exitCode, timeoutInfo = executeShCmd(^M                                                                                                                                          ^[_bk;t=1737572756104^G    exitCode, timeoutInfo = executeShCmd(^M                                                                                                                                          ^[_bk;t=1737572756104^G    exitCode, timeoutInfo = executeShCmd(^M                                                                                                                                          ^[_bk;t=1737572756104^G    exitCode, timeoutInfo = executeShCmd(^M                                                                                                                                          ^[_bk;t=1737572756104^G    exitCode, timeoutInfo = executeShCmd(^M                                                                                                                                          ^[_bk;t=1737572756104^G    res = _executeShCmd(cmd.lhs, shenv, results, timeoutHelper)^M                                                                                                                    ^[_bk;t=1737572756104^G    res = _executeShCmd(cmd.lhs, shenv, results, timeoutHelper)^M                                                                                                                    ^[_bk;t=1737572756104^G    res = _executeShCmd(cmd.lhs, shenv, results, timeoutHelper)^M                                                                                                                    ^[_bk;t=1737572756104^G    if timeoutHelper.timeoutReached():^M                                                                                                                                             ^[_bk;t=1737572756104^G    if timeoutHelper.timeoutReached():^M
```

GeorgeKA committed:
Formatting and wording changes

GeorgeKA committed:
Python darker format fixes

Maetveis commented:
> Hi @Maetveis . This is ready for review.
> 
> There are a few things that I came across while making the script that I should highlight :
> 
>     1. clang -cc1 test directory
> 
> 
> [test/Driver/lit.local.cfg](https://github.com/llvm/llvm-project/blob/e10d551aa482ee185a80216b2670a2947a8bdeb0/clang/test/Driver/lit.local.cfg#L27) says, '_("%clang_cc1", """*** Do not use 'clang -cc1' in Driver tests. ***""")_', yet I see various tests in that directory that do so, like [linker-wrapper.c](https://github.com/llvm/llvm-project/blob/e10d551aa482ee185a80216b2670a2947a8bdeb0/clang/test/Driver/linker-wrapper.c#L10). If there's no opposition, I'll leave the cc1 tests in the main Lit test file, Driver/unsupported_in_drivermode.c.

I think this should be fine, but we'll see if other reviewers disagree.

> 
>     2. flang vs flang -fc1
> 
> 
> As per [clang/test/Driver/flang/flang.f90](https://github.com/llvm/llvm-project/blob/e10d551aa482ee185a80216b2670a2947a8bdeb0/clang/test/Driver/flang/flang.f90#L1), "flang -fc1 is invoked when in --driver-mode=flang", so I disabled the plain flang tests since there's no distinction.

I don't think that comment means that `flang` and `flang -fc1` take the same options, the test is verifying that the driver (plain `flang`) calls the backend compiler (`flang -fc1`), but the supported options are not going to be the same between these two. 

> I disabled the tests on Windows given the following recursion error that occurs on Windows regression machines.

I think this might be caused by the sheer amount of `RUN:` lines in the generated test, I have an idea on how to fix it, that would also make the test faster to run. I'll post it as part of the review.

Maetveis commented on clang/utils/generate_unsupported_in_drivermode.py:
```diff
@@ -0,0 +1,626 @@
+for driver_ctrl in driver_controller:
+    driver_ctrl.supported_sequence.sort(key=len, reverse=True)
+
+# For a given driver, this script should not generate tests for unsupported options
+# whose option Name have a prefix that corresponds to a supported option / visible option of Kind KIND_JOINED*.
+# These driver-option pairs are removed here.
```
> For a given driver, this script **should** not generate tests for unsupported options
whose option Name have a prefix that corresponds to a supported option / visible option of Kind KIND_JOINED*.

Its not that it should not, instead it cannot, because that will be parsed as the prefixed option with a value and thus will not cause an error. Would be good to add a comment explaining this with a specific example.

Maetveis reviewed (changes requested):
Phew, first round in. Please don't take the number of suggestions as a negative, the basic structure looks good there's just many small tweaks to be made :).

GeorgeKA commented on clang/utils/generate_unsupported_in_drivermode.py:
```diff
@@ -0,0 +1,626 @@
+
+    # Check for the option's group's visibility
+    find_groups(group_sequence, options_json, option)
+    if len(group_sequence) > 0:
+        for group_name in group_sequence:
+            for visibility in options_json[group_name]["Visibility"]:
+                filtered_visibility.append(visibility["def"])
```
I made some changes to get_visibility in the next commit. Noting that I included a None check for what's referred here as group_sequence, (next commit group_set.) Without it, I get the following:
`TypeError: 'NoneType' object is not iterable`

GeorgeKA commented on clang/utils/generate_unsupported_in_drivermode.py:
```diff
@@ -0,0 +1,626 @@
+        return result
+
+
+def find_groups(group_sequence, options_json, option):
+    """Find the groups for a given option
+    Note that groups can themselves be part of groups, hence the recursion
+
+    group_sequence: A sequence to which group names will be appended.
+    options_json: The converted Python dictionary from the Options.td json string
+    option: The option object from Options.td
+    """
+    group_json = options_json[option]["Group"]
+
+    if group_json is None:
+        return
+
+    # Prevent circular group membership lookup
+    if len(group_sequence) > 0:
+        for group in group_sequence:
+            if group_json["def"] == group:
+                return
+
+    group_sequence.append(group_json["def"])
+    return find_groups(group_sequence, options_json, option)
```
Ah yes, there's an error here. What I intended was to recursively call find_groups with the new found group as the option so that all the relevant groups could be collected.

Good idea with the group format comment.

GeorgeKA commented on clang/utils/generate_unsupported_in_drivermode.py:
```diff
@@ -0,0 +1,626 @@
+    Note that groups can themselves be part of groups, hence the recursion
+
+    group_sequence: A sequence to which group names will be appended.
+    options_json: The converted Python dictionary from the Options.td json string
```
Yes, it's the full tablegen output.
I'll switch the name to something more appropriate.

GeorgeKA commented on clang/utils/generate_unsupported_in_drivermode.py:
```diff
@@ -0,0 +1,626 @@
+    unsupported_sequence.remove(skip_pair)
+skipped_sequence.clear()
+
+write_lit_test(LIT_TEST_PATH, VISIBILITY_DEFAULT, unsupported_sequence)
+write_lit_test(LIT_TEST_PATH_FLANG, VISIBILITY_FLANG, unsupported_sequence)
```
My thought was to follow the pre-existing conventions I saw. The only `--driver-mode=flang` usage I saw was under `clang/test/Driver/flang`. Thoughts on this?

GeorgeKA commented on clang/utils/generate_unsupported_in_drivermode.py:
```diff
@@ -0,0 +1,626 @@
+                filtered_visibility.append(visibility["def"])
+
+
+def find_supported_seq_cmp_start(supported_sequence, low, high, search_option):
```
> I think in this case it would be a binary search on the length of items?

Yup, binary search on the length of the option names.

> Also consider using a standard library function like bisect_left from the bisect module instead of rolling your own binary search.

👍

GeorgeKA commented on clang/utils/generate_unsupported_in_drivermode.py:
```diff
@@ -0,0 +1,626 @@
+        or kind == "KIND_INPUT"
+        or kind == "KIND_UNKNOWN"
+    ):
+        untested_sequence.append(
```
You're right. Removed it.
(I had it here before because I was printing which driver-option pairs weren't being checked.)

GeorgeKA commented on clang/utils/generate_unsupported_in_drivermode.py:
```diff
@@ -0,0 +1,626 @@
+                    ):
+                        continue
+
+                    # In testing, return codes cannot be relied on consistently for assessing command failure.
+                    # Leaving this handling here in case things change, but for now, Lit tests will accept pass or fail
+                    # lit_not = "not " if unsupported_pair.is_error else ""
+                    lit_not = "not not --crash "
```
I ended up removing this logic entirely. I think it was something that I carried over prior to sorting out my clang path errors. The commands all error out as expected.

GeorgeKA replied:
Nvm. Spoke too soon. Adding "Werror" was still necessary for cl commands.

GeorgeKA committed:
First review round. Switched to batch option testing

- Re-enabled flang tests.
- Renamed class DriverController to DriverData, and made it a dataclass
- Removed the custom file validation code. argparse will handle most of this
- Redefined a few lists as sets and dictionaries
- Removed untested_sequence. It was unused in the end
- Removed find_supported_seq_cmp_start() in favour of bisect_left
- Removed the command preprocessing. The problem was not the automatic addition of "cc1" & "cc1as" options. "KIND_COMMAJOINED" was omitted from the prefix check. See the comment "this script cannot generate tests for" for details.
- Switched to creating Lit test run lines with batches of test options. Improves performance, and hopefully will make the tests run successfully on Windows.

GeorgeKA commented on clang/utils/generate_unsupported_in_drivermode.py:
```diff
@@ -0,0 +1,626 @@
+                    if test_visibility == VISIBILITY_FLANG and not is_flang_driver:
+                        continue
+                    elif test_visibility == VISIBILITY_DEFAULT and is_flang_driver:
+                        continue
```
It did actually prevent flang CHECK statements from being written to the non-flang Lit test, and vice-versa, but anyway, this second check is gone in the most recent commit.

GeorgeKA commented on clang/utils/generate_unsupported_in_drivermode.py:
```diff
@@ -0,0 +1,626 @@
+
+    for alias in alias_sequence:
+        if options_json[alias]["Alias"]["def"] == option:
+            get_visibility(alias, tmp_visibility_list)
```
Yes, I assumed that was the case through observation.
That said, after taking a second look, at this point there's only one option that's affected by this logic; `_no_warnings`. 

Test syntax:
`clang --driver-mode=cl --no-warnings -### /c /WX -Werror`

`_no_warnings` has the alias `w`, which also doesn't have cl visibility. `w` also unexpectedly succeeds when tested manually, but the option is omitted from testing by the kind joined prefix processing since its prefix is `_SLASH_w`, (so still "w".)

So maybe a more accurate way to go about this would be instead of adding the visibility of an option's aliases, if an option is omitted due to the kind-joined prefix logic, also omit other options that list this option as an alias. Any qualms with that?

GeorgeKA replied:
Ah, just noticing this comment now. Noted. I'll push an update.

GeorgeKA commented:
> Phew, first round in. Please don't take the number of suggestions as a negative, the basic structure looks good there's just many small tweaks to be made :).

Thanks for all the comments @Maetveis . I worked my way through them, and this is ready for another look. The outstanding item is the alias logic. I left [a comment](https://github.com/llvm/llvm-project/pull/120900#discussion_r1938143546). I have a change ready, but I'll hold off while you respond.

GeorgeKA commented on clang/utils/generate_unsupported_in_drivermode.py:
```diff
@@ -0,0 +1,473 @@
+# For a given driver, this script cannot generate tests for unsupported options whose option "Name" have a prefix that
+# corresponds to a supported option / visible option of Kind *JOINED*. These driver-option pairs are removed here.
+# The reason is that those options will be parsed as if they were the corresponding prefixed options with a value,
+# and thus would not error would be triggered.
```
Typo here. I'll fix it in the next commit.
"_and thus would not error would be triggered._"

GeorgeKA commented:
Fixes https://github.com/llvm/llvm-project/issues/119190

GeorgeKA commented:
Fixes: https://github.com/llvm/llvm-project/issues/119190

GeorgeKA committed:
Corrected set initializations

The same Lit tests are produced as the last commit

Maetveis commented on clang/utils/generate_unsupported_in_drivermode.py:
```diff
@@ -0,0 +1,472 @@
+    lit_cmd_prefix: str
+    lit_cmd_options: str
+    visibility_str: str
+    lit_cmd_end: str = " - < /dev/null 2>&1 | FileCheck -check-prefix=CHECK-COUNT-"
+    check_str: str = "{{(unknown argument|n?N?o such file or directory)}}"
```
```suggestion
    lit_cmd_end: str = " - < /dev/null 2>&1 | FileCheck -check-prefix=CHECK-COUNT-"
    check_str: str = "{{(unknown argument|[Nn]o such file or directory)}}"
```
we wish to match either capital or lower case 'n', but shouldn't allow 'o such file or directory'.

Maetveis commented on clang/utils/generate_unsupported_in_drivermode.py:
```diff
@@ -0,0 +1,472 @@
+    " -### /c /WX -Werror",
+    VISIBILITY_CL,
+    f" 2>&1 | FileCheck -check-prefix={VISIBILITY_CL}CHECK-COUNT-",
+    "{{(unknown argument ignored in|no such file or directory|argument unused during compilation)}}",
+)
+driver_dxc = DriverData(
+    "%clang_dxc ",
+    " -### /T lib_6_7",
+    VISIBILITY_DXC,
+    f" 2>&1 | FileCheck -check-prefix={VISIBILITY_DXC}CHECK-COUNT-",
+    "{{(unknown argument|no such file or directory|argument unused during compilation)}}",
+)
+driver_default = DriverData(
+    "%clang ",
+    " -### -x c++ -c",
+    VISIBILITY_DEFAULT,
+    f" - < /dev/null 2>&1 | FileCheck -check-prefix={VISIBILITY_DEFAULT}CHECK-COUNT-",
+    "{{(unknown argument|unsupported option|argument unused|no such file or directory)}}",
+)
+driver_fc1 = DriverData(
+    "%clang --driver-mode=flang -fc1 ",
+    "",
+    VISIBILITY_FC1,
+    f" - < /dev/null 2>&1 | FileCheck -check-prefix={VISIBILITY_FC1}CHECK-COUNT-",
+    "{{(unknown argument|no such file or directory|does not exist)}}",
+)
+driver_flang = DriverData(
+    "%clang --driver-mode=flang ",
+    " -### -x c++ -c",
+    VISIBILITY_FLANG,
+    f" - < /dev/null 2>&1 | FileCheck -check-prefix={VISIBILITY_FLANG}CHECK-COUNT-",
+    "{{unknown argument|unsupported option|argument unused during compilation|invalid argument|no such file or directory}}",
```
How can we end up with the following warnings / errors?
- "no such file or directory"
- "does not exist"
- "argument unused during compilation"

I would expect that when parsing an option fails we'd basically only get "unknown argument".

GeorgeKA replied:
I've simplified the expected error strings in the latest commit. To explain "_no such file or directory_", that message is returned when an unsupported option has the "/" prefix. 

Ex:
```
clang /A  -### -x c++ -c - < /dev/null 2>&1
  clang: error: no such file or directory: '/A' 
```

Ultimately I switched to using prefix "-" instead of "/", so this is no longer an issue.

Why the switch? When testing options in batches, regardless of the order in which the options were originally provided, the error messages would be grouped by prefix used. This was causing the tests to fail. Creating handling for ordering options based on their prefix was making things messy.
Fyi, all of the "/" prefix options support "-" as well. Used this regex to double check (with gnu-grep):

`grep '"\/",\n\s6}(?!w"-")' -Poz Options.td.json`

Maetveis commented on clang/utils/generate_unsupported_in_drivermode.py:
```diff
@@ -0,0 +1,472 @@
+
+                    last_batch_size = len(batch)
+
+                # CHECK statements. Instead of writing custom CHECK statements for each RUN line, create two statements
+                # per driver. One statement for a full batch, and a second for a partial batch.
```
We can very easily run into false positives in the test this way, if a single option in the batch produces the warning we want the test will still pass even if all the rest are erroneously accepted.
It should be fine to write a check line for each warning.

GeorgeKA replied:
Yup. This was an oversight. Corrected it in latest commit. I'll upload it later today.

Maetveis reviewed (changes requested):
Thanks, this is starting to look right. I'm sorry for the long response time, I could not justify spending work time on the review this time, and my weekends have been somewhat busy.

I'll try to get some other reviewers to look at this too, to approve of the proposed workflow when updating Options.td. I would appreciate if you could make a summary of that as a comment, but if not I'll do it later.

GeorgeKA commented on clang/utils/generate_unsupported_in_drivermode.py:
```diff
@@ -0,0 +1,472 @@
+            )
+
+# Sort the supported lists for the next block
+for visibility, driver_data in driver_data_dict.items():
+    driver_data.supported_sequence.sort(key=len, reverse=True)
```
Made this change, but just noting that I needed to iterate over driver_data_dict.values().

GeorgeKA committed:
Merge branch 'llvm:main' into 119190-unsupportedOptions

GeorgeKA committed:
Round 2 of reviews

- Renamed find_groups() to collect_transitive_groups(), and simplified the function
- Introduced additional test file for flang -fc1 tests. unsupported_in_flang_fc1.f90
- Simplified Lit check statement regexes
- Created Lit check statements for each option's error messages

GeorgeKA committed:
Darker formatting change

GeorgeKA committed:
Merge branch 'llvm:main' into 119190-unsupportedOptions

GeorgeKA committed:
Small comment update

GeorgeKA commented:
@Maetveis, referring to the proposed workflow review comment, not sure if you meant to leave a comment in the script, or here in the review. I'll leave it here and in the description for now, but let me know if you mean the script.

### Proposed workflow
- A change is made to _clang/include/clang/Driver/Options.td_
- Regenerate the Lit tests that validate exposure of options to intended driver modes
`python generate_unsupported_in_drivermode.py --llvm-bin <path>/llvm-project/build/bin --llvm-tblgen llvm-tblgen ../include/clang/Driver/Options.td`
- Run the Lit tests to check exposure                                                                                                                                                                           
`llvm-lit -v ~/llvm-project/clang/test/Driver/unsupported_in_drivermode.c`
`llvm-lit -v ~/llvm-project/clang/test/Driver/flang/unsupported_in_drivermode.c`
`llvm-lit -v ~/llvm-project/flang/test/Driver/unsupported_in_flang_fc1.f90`

GeorgeKA committed:
Removed the alias handling

Also added _no_warnings to the exception sequence

GeorgeKA committed:
Small comment change

GeorgeKA commented:
- I should draw attention to a change I made when it comes to option prefixes. Put simply, I had to switch to using prefix "-" instead of "/" because of different error output for "/". I describe it [here](https://github.com/llvm/llvm-project/blob/1e6150bc23c3410945cddb9d1d33f6e57f5c3f60/clang/utils/generate_unsupported_in_drivermode.py#L404).

- Secondly, I had to add handling for matching option names, which was distinguished before by differing prefixes. Described [here](https://github.com/llvm/llvm-project/blob/1e6150bc23c3410945cddb9d1d33f6e57f5c3f60/clang/utils/generate_unsupported_in_drivermode.py#L450)

- Also, I moved the flang fc1 testing to _flang/test/Driver/unsupported_in_flang_fc1.f90_ in order to use `%flang_fc1`

Maetveis commented:
Hey @GeorgeKA, there's still some unaddressed comments from my previous review, I thought you were still working on it, so I did not want to post more comments. I realize the latency I'm taking to respond is dragging this out quite a bit. If you're okay with it I can implement some of them myself. Github allows to do this by opening a pull request against this branch in your fork of the repository.

GeorgeKA commented:
Hi @Maetveis.

> there's still some unaddressed comments from my previous review, I thought you were still working on it

Oh my mistake. I didn't get a notification, so I didn't see the most recent comments.

> Also it seems like the good first issue label might have been inaccurate, I really didn't expect so much complexity.

Don't sweat it. I appreciate the opportunity.

> I'd be happy to help with implementing some of my suggestions if you are okay with that. Github allows to do this by opening a pull request against this branch in your fork of the repository.

Fine by me.

GeorgeKA committed:
Changed argument parsing. UnsupportedDriverOption -> DriverOption

Changed UnsupportedDriverOption to DriverOption, and converted to
dataclass.
Adjusted argument parsing

GeorgeKA commented on clang/utils/generate_unsupported_in_drivermode.py:
```diff
@@ -0,0 +1,472 @@
+    [
+        tablegen,
+        "-I",
+        os.path.join(os.path.dirname(__file__), "../../llvm/include"),
```
The include directory for llvm-tblgen is different than where Options.td is located, so I added an extra path option just now, `--options-td-dir`. Not sure if that's what you were going for, so let me know if you'd like me to change it (or feel free to make the edit yourself like you were asking.)

GeorgeKA commented on clang/utils/generate_unsupported_in_drivermode.py:
```diff
@@ -0,0 +1,626 @@
+    unsupported_list: List of UnsupportedDriverOption objects
+    """
+    try:
+        with open(test_path, "w") as lit_file:
```
Just removed it

GeorgeKA commented on clang/utils/generate_unsupported_in_drivermode.py:
```diff
@@ -0,0 +1,472 @@
+
+    for supported_pair in supported_seq[start_index:]:
+        if (
+            unsupported_pair.option_name.startswith(supported_pair.option_name)
```
To avoid confusion, I'm going to refer to option name string prefixes as "option-name-prefixes", and the delimiting character for options as an "option-prefix."
option-name-prefix: "A" is an option-name-prefix for "AI"
option-prefix: "/" or "-"

I just tested the following which suggests that we should only check for option-name-prefixes and disregard matching option-prefixes.

I added handling for checking if the unsupported pair primary option-prefix is in the option-prefix list for a supported pair. These tests suggest that regardless of the option-prefix, if there's a matching option-name-prefix for a JOINED option, the JOINED option will be assumed.

For reference, this is the output when a supported CC1AS option is used:
```
$clang -cc1as  -compress-debug-sections -
clang -cc1as: error: unknown target triple 'unknown'
```

1.
Option "I", option name "I", prefixes ["-"], has visibility CC1AsOption
Option "_SLASH_I", option name "I", prefixes ["/", "-"], doesn't have visibility CC1AsOption


```
$ clang -cc1as  -I  -
clang -cc1as: error: unknown target triple 'unknown'
```

2.
None of these options have visibility CC1AS. All have prefixes ["/", "-"] 

```
$clang -cc1as  -openmp -openmp- -openmp:experimental -
clang -cc1as: error: unknown target triple 'unknown'
```

In this case, I'm assuming the option-name-prefix that's being used is "o" for options "o" or "_SLASH_o".

3.
Option "D", option name "D", prefixes ["-"], has visibility CC1Option
Option "_SLASH_D", option name "D", prefixes ["/", "-"], doesn't have visibility CC1Option

```
$clang -cc1 -D -x c++ -
error: error reading 'c++': No such file or directory
```

Maetveis replied:
Okay, basically because dash (`-`) is always used, and supported for all options the checks would be redundant.

GeorgeKA committed:
Lit test formatting change & optional path arguments

Switched to writting each option on its own line for easier git diffs.
Added options for Lit test file paths.

Maetveis committed:
Change Options File Dir to Options File Path

No reason to assume file will be named Options.td

Maetveis commented on clang/utils/generate_unsupported_in_drivermode.py:
```diff
@@ -0,0 +1,534 @@
+
+        comment_str = EXCLAMATION if is_flang_pair else SLASH_SLASH
+
+        unflattened_option_data = list(
+            itertools.batched(driver_data.test_option_sequence, batch_size)
+        )
+
+        for i, batch in enumerate(unflattened_option_data):
```
```suggestion
        batched = itertools.batched(driver_data.test_option_sequence, batch_size)
        for i, batch in enumerate(batched):
```
No need to collapse the iterable to a list.
I don't think the variable name needs to be that verbose.

Maetveis commented on clang/utils/generate_unsupported_in_drivermode.py:
```diff
@@ -0,0 +1,534 @@
+        ):
+            continue
+
+        comment_str = EXCLAMATION if is_flang_pair else SLASH_SLASH
```
```suggestion
        comment_str = "! " if is_flang_pair else "// "
```
And remove `SLASH_SLASH` and `EXCLAMATION`

Maetveis commented on clang/utils/generate_unsupported_in_drivermode.py:
```diff
@@ -0,0 +1,534 @@
+driver_cc1 = DriverData(
+    "%clang -cc1 ",
+    " -x c++",
+    VISIBILITY_CC1,
+    f" - < /dev/null 2>&1 | FileCheck -check-prefix={VISIBILITY_CC1}CHECK",
+    "unknown argument",
+)
+driver_cl = DriverData(
+    "%clang_cl ",
+    " -### /c /WX -Werror",
+    VISIBILITY_CL,
+    f" 2>&1 | FileCheck -check-prefix={VISIBILITY_CL}CHECK",
+    "unknown argument ignored in clang-cl",
+)
+driver_dxc = DriverData(
+    "%clang_dxc ",
+    " -### /T lib_6_7",
+    VISIBILITY_DXC,
+    f" 2>&1 | FileCheck -check-prefix={VISIBILITY_DXC}CHECK",
+    "unknown argument",
+)
+driver_default = DriverData(
+    "%clang ",
+    " -### -x c++ -c",
+    VISIBILITY_DEFAULT,
+    f" - < /dev/null 2>&1 | FileCheck -check-prefix={VISIBILITY_DEFAULT}CHECK",
+    "unknown argument",
+)
+driver_fc1 = DriverData(
+    "%flang_fc1 ",
+    "",
+    VISIBILITY_FC1,
+    f" - < /dev/null 2>&1 | FileCheck -check-prefix={VISIBILITY_FC1}CHECK",
+    "unknown argument",
+)
+driver_flang = DriverData(
+    "%clang --driver-mode=flang ",
+    " -### -x c++ -c",
+    VISIBILITY_FLANG,
+    f" - < /dev/null 2>&1 | FileCheck -check-prefix={VISIBILITY_FLANG}CHECK",
```
```suggestion
    VISIBILITY_CC1AS,
    f" - < /dev/null 2>&1 | FileCheck -check-prefix=CHECK_CC1AS_",
    "unknown argument",
)
driver_cc1 = DriverData(
    "%clang -cc1 ",
    " -x c++",
    VISIBILITY_CC1,
    f" - < /dev/null 2>&1 | FileCheck -check-prefix=CHECK_CC1_",
    "unknown argument",
)
driver_cl = DriverData(
    "%clang_cl ",
    " -### /c /WX -Werror",
    VISIBILITY_CL,
    f" 2>&1 | FileCheck -check-prefix=CHECK_CL_",
    "unknown argument ignored in clang-cl",
)
driver_dxc = DriverData(
    "%clang_dxc ",
    " -### /T lib_6_7",
    VISIBILITY_DXC,
    f" 2>&1 | FileCheck -check-prefix=CHECK_DXC_",
    "unknown argument",
)
driver_default = DriverData(
    "%clang ",
    " -### -x c++ -c",
    VISIBILITY_DEFAULT,
    f" - < /dev/null 2>&1 | FileCheck -check-prefix=CHECK_CLANG_",
    "unknown argument",
)
driver_fc1 = DriverData(
    "%flang_fc1 ",
    "",
    VISIBILITY_FC1,
    f" - < /dev/null 2>&1 | FileCheck -check-prefix=CHECK_FC1_",
    "unknown argument",
)
driver_flang = DriverData(
    "%clang --driver-mode=flang ",
    " -### -x c++ -c",
    VISIBILITY_FLANG,
    f" - < /dev/null 2>&1 | FileCheck -check-prefix=CHECK_FLANG_",
```

I think these are more fitting than `FC1OptionCHECK`.

GeorgeKA replied:
Gotcha. I had to introduce `check_prefix` to DriverData so that I could reference these new strings in `write_lit_test()` for the CHECK lines.
The strings are duplicated when creating the DriverData objects, but I figure this was another readability situation.

Maetveis commented on clang/utils/generate_unsupported_in_drivermode.py:
```diff
@@ -0,0 +1,534 @@
+    test_visibility: Any VISIBILITY_* variable. VISIBILITY_DEFAULT will return the .c formatted test note.
+    All other will return the .f90 formatted test note
+    """
+    test_prefix = SLASH_SLASH if test_visibility == VISIBILITY_DEFAULT else EXCLAMATION
```
I guess here too, use ! and // instead of the constants. It's not a problem that it's "duplicated", as it makes reading the code much simpler.

Maetveis commented on clang/utils/generate_unsupported_in_drivermode.py:
```diff
@@ -0,0 +1,534 @@
+            visibility_set.add(visibility["def"])
+
+    # Check for the option's group's visibility
+    group_set = collect_transitive_groups(option, options_dictionary)
+    if group_set:
+        for group_name in group_set:
+            for visibility in options_dictionary[group_name]["Visibility"]:
+                visibility_set.add(visibility["def"])
```
```suggestion
     for group_name in collect_transitive_groups(option, options_dictionary):
         for visibility in options_dictionary[group_name]["Visibility"]:
             visibility_set.add(visibility["def"])
```
`collect_transitive_groups` shouldn't ever return `None` now.

Maetveis commented on clang/utils/generate_unsupported_in_drivermode.py:
```diff
@@ -0,0 +1,534 @@
+                    + ": {{("
+                    + driver_data.check_str  # "unknown argument"
+                    + ").*"
+                    + option_str.replace("+", "\\+")  # "-A"
```
```suggestion
                    + re.escape(option_str)  # "-A"
```
This really should be minimally [re.escape](https://docs.python.org/3/library/re.html#re.escape), but it could be taken out of the regex too i.e.
`unknown argument{{.*}}-A`

GeorgeKA committed:
Merge branch 'llvm:main' into 119190-unsupportedOptions

GeorgeKA committed:
Merge pull request #1 from Maetveis/options_file

Change Options File Dir to Options File Path

GeorgeKA committed:
Changed check-prefix strings for readability

GeorgeKA commented:
Ping @Maetveis 
Any more thoughts on this one? I think I've addressed the existing comments.

GeorgeKA committed:
Merge branch 'llvm:main' into 119190-unsupportedOptions

GeorgeKA committed:
Merge branch 'llvm:main' into 119190-unsupportedOptions

GeorgeKA committed:
Refreshing Lit tests

GeorgeKA commented on clang/utils/generate_unsupported_in_drivermode.py:
```diff
@@ -0,0 +1,472 @@
+    help="LLVM TableGen executable. If not included with --llvm-bin, the script will search for the llvm-tblgen executable",
+)
+
+args = vars(parser.parse_args())
```
Updated this in a previous commit.

GeorgeKA commented on clang/utils/generate_unsupported_in_drivermode.py:
```diff
@@ -0,0 +1,472 @@
+    prefix = ""
+    if prefixes is not None and len(prefixes) > 0:
+        # Assuming the first prefix is the preferred prefix
+        prefix = prefixes[0]
```
A change to deal with misleading output from "/" prefixed options covers this. [This comment](https://github.com/llvm/llvm-project/blob/66e62df1780af56947f669905aab3eae0eee679f/clang/utils/generate_unsupported_in_drivermode.py#L490) explains it.

GeorgeKA committed:
Moved preceding and trailing white space DriverData instantiations to write_lit_test

GeorgeKA committed:
Merge branch 'llvm:main' into 119190-unsupportedOptions

GeorgeKA committed:
Updating Lit tests
